### ISSUE TITLE
ETL CDMBuilder - Errors in 'Conversion from RAW to CDM' step

### ISSUE THREAD

**Issue by pkumar81** [OHDSI/ETL-CDMBuilder#21]
We are converting Truven MDCR data (2003-2015) to CDM format using ETL CDMBuilder. In 'Conversion from RAW to CDM' step, we got these errors:
1. Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.
2. Message: This SqlTransaction has completed; it is no longer usable.

After first error, ETL CDMBuilder didn't stop and it kept processing the data. But it has stopped processing any data after the second error.  'Conversion from RAW to CDM' step has been running for hours to process the last two batches (19123 from 19124). I looked into programs Chunkbuilder.cs and Saver.cs, but didn't get much information about these errors. Could you please suggest me what should be the next step so that ETL CDMBuilder can run successfully.

The details of the errors are as follows:

Builder: HSC-CTSC-TRUVEN
Time: Mar  3 2017  5:09PM
Error: Exception Found:
Type: System.AggregateException
Message: One or more errors occurred.
Source: mscorlib
Stacktrace:    at System.Threading.Tasks.Task.WaitAll(Task[] tasks, Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at org.ohdsi.cdm.presentation.builderprocess.Program.Main(String[] inputArgs)
  Inner Exception Found:
  Type: System.InvalidOperationException
  Message: This SqlTransaction has completed; it is no longer usable.
  Source: System.Data
  Stacktrace:    at System.Data.SqlClient.SqlTransaction.ZombieCheck()
   at System.Data.SqlClient.SqlTransaction.Rollback()
   at org.ohdsi.cdm.framework.core.Savers.Saver.SaveV5(ChunkData chunk, Boolean lastSave)
   at org.ohdsi.cdm.framework.core.Base.ChunkBuilder.SaveData1()
   at System.Threading.Tasks.Task.Execute()

Builder: HSC-CTSC-TRUVEN
Time: Mar  2 2017  9:26PM
Error: Exception Found:
Type: System.Data.SqlClient.SqlException
Message: Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.
Source: .Net SqlClient Data Provider
Stacktrace:    at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParserStateObject.ReadSniError(TdsParserStateObject stateObj, UInt32 error)
   at System.Data.SqlClient.TdsParserStateObject.ReadSniSyncOverAsync()
   at System.Data.SqlClient.TdsParserStateObject.TryReadNetworkPacket()
   at System.Data.SqlClient.TdsParserStateObject.TryPrepareBuffer()
   at System.Data.SqlClient.TdsParserStateObject.TryReadByte(Byte& value)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.TdsParser.Run(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj)
   at System.Data.SqlClient.TdsParser.TdsExecuteTransactionManagerRequest(Byte[] buffer, TransactionManagerRequestType request, String transactionName, TransactionManagerIsolationLevel isoLevel, Int32 timeout, SqlInternalTransaction transaction, TdsParserStateObject stateObj, Boolean isDelegateControlRequest)
   at System.Data.SqlClient.SqlInternalConnectionTds.ExecuteTransactionYukon(TransactionRequest transactionRequest, String transactionName, IsolationLevel iso, SqlInternalTransaction internalTransaction, Boolean isDelegateControlRequest)
   at System.Data.SqlClient.SqlInternalTransaction.Commit()
   at System.Data.SqlClient.SqlTransaction.Commit()
   at org.ohdsi.cdm.framework.core.Savers.MSSqlSaver.Commit()
   at org.ohdsi.cdm.framework.core.Savers.Saver.SaveV5(ChunkData chunk, Boolean lastSave)
  Inner Exception Found:
  Type: System.ComponentModel.Win32Exception
  Message: The wait operation timed out
  Source: 
  Stacktrace:

**Comment by pkumar81** [comment]
I looked into the log table under Builder database and found that we have all the chunk ids present in the log table. Also, there is a row with chunkid 'NULL' and the message column of that NULL chunkid has the same error message:

Id	ChunkId	BuilderId	Type	Time	Message	BuildingId
554598	NULL	1	Error	2017-03-03 17:09:23.583	Exception Found:
Type: System.AggregateException
Message: One or more errors occurred.
Source: mscorlib
Stacktrace:    at System.Threading.Tasks.Task.WaitAll(Task[] tasks, Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at org.ohdsi.cdm.presentation.builderprocess.Program.Main(String[] inputArgs)
  Inner Exception Found:
  Type: System.InvalidOperationException
  Message: This SqlTransaction has completed; it is no longer usable.
  Source: System.Data
  Stacktrace:    at System.Data.SqlClient.SqlTransaction.ZombieCheck()
   at System.Data.SqlClient.SqlTransaction.Rollback()
   at org.ohdsi.cdm.framework.core.Savers.Saver.SaveV5(ChunkData chunk, Boolean lastSave)
   at org.ohdsi.cdm.framework.core.Base.ChunkBuilder.SaveData1()
   at System.Threading.Tasks.Task.Execute()	1

**Comment by ChristopheLambert** [comment]
Praveen was able to get these errors to go away by increasing the timeout time for queries. Praveen, can you share the steps you did and close this out?

**Comment by pkumar81** [comment]
To fix the problems, I did the following two things:
1.) For the first issue (Execution Timeout Expired):  In the configuration file 'org.ohdsi.cdm.presentation.buildingmanager.exe.Config', we added Connect Timeout=3000; to Source, Destination, and Vocabulary in connectionStrings tag.

2.) For the second issue (This SqlTransaction has completed; it is no longer usable): Our machine, where SQL Server is installed, had only 8 cores and we had set MaxdegreeofParallism to 8. This caused the second problem. To fix it, we ran the ETL-CDMBuilder with MaxdegreeofParallism=4 and Batch Size = 512 (2**n)
